src/Lean/Meta/IndPredBelow.lean

```diff
@@ -374,6 +378,183 @@ where
       forallBoundedTelescope rest (some 1) fun ys rest =>
       loop xs rest belowIndices xIdx (yIdx + 1)
 
+private def belowType (motive : Expr) (xs : Array Expr) (idx : Nat) : MetaM $ Name × Expr := do
+  (←inferType xs[idx]).withApp fun type args => do
+  let indName := type.constName!
+  let indInfo ← getConstInfoInduct indName
+  let belowArgs := args[:indInfo.numParams] ++ #[motive] ++ args[indInfo.numParams:] ++ #[xs[idx]]
+  let belowType := mkAppN (mkConst (indName ++ `below) type.constLevels!) belowArgs
+  (indName, belowType)
+
+partial def mkBelowMatcher
+    (matcherApp : MatcherApp) 
+    (belowMotive : Expr) 
+    (below : Expr)
+    (idx : Nat) : MetaM $ Expr × MetaM Unit := do
+  let mkMatcherInput ← getMkMatcherInputInContext matcherApp
+  let (indName, belowType, motive, matchType) ← 
+    forallBoundedTelescope mkMatcherInput.matchType mkMatcherInput.numDiscrs fun xs t => do
+    let (indName, belowType) ← belowType belowMotive xs idx
+    let matchType ← 
+      withLocalDeclD (←mkFreshUserName `h_below) belowType fun h_below => do
+      mkForallFVars (xs.push h_below) t
+    let motive ← newMotive belowType xs
+    (indName, belowType.replaceFVars xs matcherApp.discrs, motive, matchType)
+
+  let lhss ← mkMatcherInput.lhss.mapM $ addBelowPattern indName 
+  let alts ← mkMatcherInput.lhss.zip lhss |>.toArray.zip matcherApp.alts |>.mapIdxM fun idx ((oldLhs, lhs), alt) => do
+    withExistingLocalDecls (oldLhs.fvarDecls ++ lhs.fvarDecls) do
+    lambdaTelescope alt fun xs t => do
+    let oldFVars := oldLhs.fvarDecls.toArray
+    let fvars := lhs.fvarDecls.toArray.map (·.toExpr)
+    let xs :=
+      -- special case: if we had no free vars, i.e. there was a unit added and no we do have free vars, we get rid of the unit.
+      match oldFVars.size, fvars.size with
+      | 0, n+1 => xs[1:]
+      | _, _ => xs
+    let t := t.replaceFVars xs[:oldFVars.size] fvars[:oldFVars.size]
+    trace[Meta.IndPredBelow.match] "xs = {xs}; oldFVars = {oldFVars.map (·.toExpr)}; fvars = {fvars}; new = {fvars[:oldFVars.size] ++ xs[oldFVars.size:] ++ fvars[oldFVars.size:]}"
+    let newAlt ← mkLambdaFVars (fvars[:oldFVars.size] ++ xs[oldFVars.size:] ++ fvars[oldFVars.size:]) t
+    trace[Meta.IndPredBelow.match] "alt {idx}:\n{alt} ↦ {newAlt}"
+    newAlt
+
+  let matcherName ← mkFreshUserName mkMatcherInput.matcherName
+  withExistingLocalDecls (lhss.foldl (init := []) fun s v => s ++ v.fvarDecls) do
+    for lhs in lhss do
+      trace[Meta.IndPredBelow.match] "{←lhs.patterns.mapM (·.toMessageData)}"
+  let res ← Match.mkMatcher { matcherName, matchType, numDiscrs := (mkMatcherInput.numDiscrs + 1), lhss }
+  res.addMatcher
+  check res.matcher
```

Comment:
Is the `check` used only for debugging? If yes, let's add a comment.
I am saying that because some constructions we have may create type incorrect terms and we use `check` to ensure they "worked", and it is useful to distinguish the two cases when maintaining the code later.

Reply:
I think it's the former kind. `Match.mkMatcher` can create a type-incorrect term. And in fact, if you are trying to do induction on the wrong argument, it does trigger. This way the error can propagate and we try a different argument. At least that's how I remember it.

If you want me to, I can try removing it, and see if the error is caught higher up, too.

Reply:
Let's add a comment with the explanation above.